video/out/hwdec/hwdec_drmprime.c

```diff
@@ -114,45 +160,44 @@ static int init(struct ra_hwdec *hw)
                               opt_path ? opt_path : "/dev/dri/renderD128";
     MP_VERBOSE(hw, "Using DRM device: %s\n", device_path);
 
-    int ret = av_hwdevice_ctx_create(&p->hwctx.av_device_ref,
-                                     AV_HWDEVICE_TYPE_DRM,
-                                     device_path, NULL, 0);
+    ret = av_hwdevice_ctx_create(&p->hwctx.av_device_ref,
+                                 AV_HWDEVICE_TYPE_DRM,
+                                 device_path, NULL, 0);
     talloc_free(tmp);
     if (ret != 0) {
         MP_VERBOSE(hw, "Failed to create hwdevice_ctx: %s\n", av_err2str(ret));
         return -1;
     }
 
-    /*
-     * At the moment, there is no way to discover compatible formats
-     * from the hwdevice_ctx, and in fact the ffmpeg hwaccels hard-code
-     * formats too, so we're not missing out on anything.
-     */
-    int num_formats = 0;
-    MP_TARRAY_APPEND(p, p->formats, num_formats, IMGFMT_NV12);
-    MP_TARRAY_APPEND(p, p->formats, num_formats, IMGFMT_420P);
-    MP_TARRAY_APPEND(p, p->formats, num_formats, pixfmt2imgfmt(AV_PIX_FMT_NV16));
-    MP_TARRAY_APPEND(p, p->formats, num_formats, IMGFMT_P010);
-#ifdef AV_PIX_FMT_P210
-    MP_TARRAY_APPEND(p, p->formats, num_formats, pixfmt2imgfmt(AV_PIX_FMT_P210));
-#endif
+    return post_init(hw);
+}
 
-    for (int i = 0; i < MP_ARRAY_SIZE(forked_pix_fmt_names); i++) {
-        enum AVPixelFormat fmt = av_get_pix_fmt(forked_pix_fmt_names[i]);
-        if (fmt != AV_PIX_FMT_NONE) {
-            MP_TARRAY_APPEND(p, p->formats, num_formats, pixfmt2imgfmt(fmt));
-        }
-    }
+#if HAVE_V4L2REQUEST
+static int init_v4l2request(struct ra_hwdec *hw)
+{
+    struct priv_owner *p = hw->priv;
 
-    MP_TARRAY_APPEND(p, p->formats, num_formats, 0); // terminate it
+    int ret = pre_init(hw);
+    if (ret != 0)
+        return ret;
 
-    p->hwctx.hw_imgfmt = IMGFMT_DRMPRIME;
-    p->hwctx.supported_formats = p->formats;
-    p->hwctx.driver_name = hw->driver->name;
-    hwdec_devices_add(hw->devs, &p->hwctx);
+    /*
+     * AVCodecHWConfig contains a combo of a pixel format and hwdevice type,
+     * correct type must be created here or hwaccel will fail.
+     *
+     * FIXME: Create hwdevice based on type in AVCodecHWConfig
+     */
+    ret = av_hwdevice_ctx_create(&p->hwctx.av_device_ref,
+                                 AV_HWDEVICE_TYPE_V4L2REQUEST,
+                                 NULL, NULL, 0);
+    if (ret != 0) {
```

Comment:
It is valid, but I think idiomatic check for ffmpeg errors is `ret < 0` generally.